### DISCUSSION TITLE
Multiple calls to build method when  reading an AutoDisposeNotifier

### DISCUSSION THREAD

**Discussion by MisterLight** [rrousselGit/riverpod#2436] [Q&A] [answered]
I have been playing around with the new Notifier/AsyncNotifier providers and replacing my old StateProviders. 

Currently I have some problems with a Notifier class that inherit from AutoDisposeNotifier. Every time I read the NotifierProvider associate to this class the build method is called. For example:

My provider with the notifier class

```
class Counter extends AutoDisposeNotifier<int> {
  @override
  int build() {
    return 0;
  }

  void increment() {
    state++;
  }
}

final counterProvider = NotifierProvider.autoDispose<Counter, int>(Counter.new);

```  

Lets say I have Widget A, with two child widget, B and C. 

```
class A extends ConsumerWidget {
  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final counter = ref.read(counterProvider);
    return Padding(
      padding: const EdgeInsets.symmetric(horizontal: widthStandard),
      child: Column(children: [
       B(),
       C(),
       CustomButton(onClick {
                final counter = ref.read(counterProvider);
               debugprint(counter.toString());
       })
  ]) 
);     
}

class B extends ConsumerWidget {
  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final counter = ref.read(counterProvider);
    return Column(children: [
      return Text("B widget'),
      ChildButton(onClick: () {
                 ref.read(counterProvider.notifier).increment();
   });
]);
}


class C extends ConsumerWidget {
  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final counter = ref.read(counterProvider);
       return Column(children: [
      return Text("C widget'),
      ChildButton(onClick: () {
                 ref.read(counterProvider.notifier).increment();
   });
]);
}
```

When the children component are builded the build method inside provider is called, setting the state to 0. Thats mean the build method inside Counter is called 2 time, one per component build. 

I can increment this value latter clicking on B and C buttons, but when I click on button inside parent widget (Widget A), and read the provider, the build method of Counter notifier is called, setting the state again to 0 and overriding the incremented value.

Is only happening when I extend  the class from AutoDisposeNotifier, but not when use Notifier.

**Comment by MisterLight** [comment]
Anyone know why is happening? is this a bug? Why  each time I read an AutoDisposeNotifier triggers the build method?

**Reply by SamNdirangu** [reply]
I also have a similar issue.

**Comment by AhmedLSayed9** [comment, marked as answer]
Don't use `ref.read` to subscribe for changes inside build. instead, use `ref.watch`.

**Comment by janoschp** [comment]
ok, but **why?**

**Reply by gdurandrexel** [reply]
An autodisposed provider is kept alive only while it is watched or listened to. If you call read(), riverpod can dispose the provider any time after the call. It surely happens when an async gap occurs, but there are probably other occurences.
If your widget depends on the value of a provider, use watch or listen.
You can find more info on the riverpod site.
